getAutocompleterUI: don't redefine ListBox component on every render

jsnajdr opened the pull request:
Fixes #50562. The `AutocompleterUI` component defined a nested `ListBox` component inside its render function, but this is an antipattern! The `ListBox` component has different identity on every render, and that causes it to completely unmount and remount on every render.

One of the symptoms is that the `onClick` handler on the `Button` subcomponents in the `ListBox` is never called when the autocompleter is inside the table. Don't ask me why 🙂 There is probably some chaotic unstable rerendering that destroys the `onClick` handler before it gets a chance to be called.

Another symptom that you can easily reproduce is that the gravatar image on the user autocomplete is blinking as you type. Because the items are not rendered in a stable way, the `img` element gets repeatedly destroyed and created and loaded again.

**How to test:**
Verify that the symptoms reported in #50562 (user autocomplete inside table cell) are fixed. Also, the gravatar images should be stabilized.

jsnajdr committed:
getAutocompleterUI: don't redefine ListBox component on every render

ellatrix reviewed (approved):
Wouldn't be bad if it also adds an e2e test to the existing suite of autocomplete tests

ellatrix commented on packages/components/src/autocomplete/autocompleter-ui.tsx:
```diff
@@ -27,11 +27,59 @@ import { VisuallyHidden } from '../visually-hidden';
+				) ) }
+			</Component>
+		);
+	}
```
Do we need it to be inside of `getAutoCompleterUI`?

jsnajdr replied:
It can be outside! Moved 🙂

alexstine reviewed (approved):
Works fine for me. Also fixes this issue: https://github.com/WordPress/gutenberg/issues/55299.

jsnajdr committed:
Move ListBox to top level

jsnajdr committed:
Add components changelog entry

jsnajdr committed:
Add e2e test for autocomplete inside table

jsnajdr commented:
> Wouldn't be bad if it also adds an e2e test to the existing suite of autocomplete tests

Good idea, added a test that autocompletes inside a table. Verified that it fails on trunk and succeeds after the fix.

I think I addressed all feedback, going to merge now.

tyxla reviewed (approved):
Really nice one, @jsnajdr 🙌
